### ISSUE TITLE
OOPS, apos.docs.insert / apos.pieces.insert / update have no options arg, no way to say "skip permissions"

### ISSUE THREAD

**Issue by boutell** [punkave/apostrophe#642]
Putting this in github, we need to be discussing A2 issues in github, not hidden away in slack. Please do respond here and not in slack.

Oh geez. This is a big oversight. Set of oversights. Constellation of oversights. Murder of crows.

I'm glad I decided to push through one more tutorial before publishing 2.0.0!

I was working on a form submission tutorial, and decided the forms would wind up being inserted as pieces. @agilbert and I have been talking about it anyway for a client project.

I was just about done when I discovered there's no way to say this in a pieces module:

```javascript
self.insert(req, { my piece ... }, { permissions: false }, callback);
```

You remember this from 0.5. It's how you insert or update something when you decide, as the programmer, that it's OK to let this anonymous person do it regardless of permissions... like submitting a public form and inserting it as a piece.

But 0.6 has no options argument to add "permissions: false" to. Oooooops.

Right now I could write this:

```javascript
self.insert(apos.tasks.getReq(), { my piece ... }, callback);
```

But that's terrible. Task reqs are for writing tasks and other times when you don't care who the user is at all. But there are plenty of use cases for making exceptions like this while still remembering who people are in our version history, etc. (Not an issue for form submissions, but could be for other cases)

In general I just don't think having no options argument for something as fundamental as `insert` and `update` is sustainable.

No, I think we need to change `docs.insert` and `pieces.insert` to take an optional `options` object. "Update" too. Just like 0.5.

The object can be optional though. Fine to call without it. So what's the hard bit?

There's a whole slew of `callAll` methods that get invoked by this stuff:

docBeforeInsert
docBeforeUpdate
docBeforeSave
docAfterUpdate
docAfterSave
docAfterInsert
docAfterDenormalizePermissions

And in the `pieces` module too, there are lots of override hooks:

beforeCreate <-- named wrong, should be beforeInsert, the method is insert!
afterCreate <-- named wrong, see above
afterConvert
beforeSave
afterSave
beforeUpdate
afterUpdate

*And* those override hooks are called only from the form routes, they should be called from the insert and update methods of the module, c'mannnn

My plan here is to implement the missing options object and the other fixes mentioned above. Then do some extensive searching to see how they will impact our projects, which might or might not have overrides that need the change to accept the options object.

Feedback encouraged. Gonna be a big part of my Friday.

cc @agilbert @colpanik @stuartromanek @kyjoya @austinstarin @grdunn

**Comment by boutell** [comment]
I did a fair amount of work on it so far this morning, and I have unit tests passing but need to add some tests that actually pass a `{ permissions: false }` object. 

This turned out to be a good opportunity to give the "write" side of our model API proper attention before shipping 2.0.0. I also stumbled over a significant oversight in our permissions checks, the fix for which will be included in my PR.

I have done searches in lansdowne, urban and sci center and I have found no overrides for any of the extension points like "afterInsert". With that in mind, it seems clear that adding the options object so that these APIs have flexibility for the future and `{permissions: false}` today is the smart way to go.

**Comment by boutell** [comment]
See:

https://github.com/punkave/apostrophe/tree/insert-and-update-fixes

I am mindful that support team's billable hours have been few this week, so I am going to switch gears and and hit that for a while, allowing y'all a chance to review this issue and the above branch.

I want to do some more testing before I merge it.

There is a small matching change in a branch of the same name for apostrophe-events.

**Comment by boutell** [comment]
apostrophe-places also got a matching change. apostrophe-blog does not need one.

**Comment by agilbert** [comment]
An optional options object makes sense to me. In 0.5, does the options parameter get used to pass anything else to self.insert?

**Comment by boutell** [comment]
0.5 uses it for some workflow-related information passing, which seems highly likely to come up again.

**Comment by boutell** [comment]
This was resolved by #645.

YOU NEED TO CHANGE YOUR CODE for any of these if you have overrides for them (SERVER SIDE only). All of them now expect an `options` argument before the `callback`.

YOU MIGHT HAVE:

beforeCreate --> NOW beforeInsert, plus the options arg before callback
afterCreate --> NOW afterInsert, plus the options arg before callback
beforeSave
afterSave
beforeUpdate
afterUpdate

NOTE: afterConvert is NOT affected.

YOU PROBABLY DON'T HAVE, BUT IF YOU DO...

docBeforeInsert
docBeforeUpdate
docBeforeSave
docAfterUpdate
docAfterSave
docAfterInsert
docAfterDenormalizePermissions
